src/Features/CSharp/Portable/QuickInfo/CSharpDiagnosticAnalyzerQuickInfoProvider.cs

```diff
@@ -0,0 +1,145 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.QuickInfo;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.CSharp.QuickInfo
+{
+    [ExportQuickInfoProvider(QuickInfoProviderNames.DiagnosticAnalyzer, LanguageNames.CSharp), Shared]
+    [ExtensionOrder(After = QuickInfoProviderNames.Syntactic)]
+    internal class CSharpDiagnosticAnalyzerQuickInfoProvider : CommonQuickInfoProvider
+    {
+        private readonly IDiagnosticAnalyzerService _diagnosticAnalyzerService;
+
+        [ImportingConstructor]
+        [SuppressMessage("RoslynDiagnosticsReliability", "RS0033:Importing constructor should be [Obsolete]", Justification = "Used in test code: https://github.com/dotnet/roslyn/issues/42814")]
+        public CSharpDiagnosticAnalyzerQuickInfoProvider(IDiagnosticAnalyzerService diagnosticAnalyzerService)
+        {
+            _diagnosticAnalyzerService = diagnosticAnalyzerService;
+        }
+
+        protected override async Task<QuickInfoItem?> BuildQuickInfoAsync(
+            Document document,
+            SyntaxToken token,
+            CancellationToken cancellationToken)
+        {
+            var (pragmaWarning, errorCode) = token.Parent switch
+            {
+                PragmaWarningDirectiveTriviaSyntax directive when directive.IsDisablePragma()
+                    => (directive, token.IsKind(SyntaxKind.EndOfDirectiveToken)
+                        ? directive.ErrorCodes.LastOrDefault() as IdentifierNameSyntax
+                        : directive.ErrorCodes.FirstOrDefault() as IdentifierNameSyntax),
+                IdentifierNameSyntax { Parent: PragmaWarningDirectiveTriviaSyntax directive } identifier when directive.IsDisablePragma()
+                    => (directive, identifier),
+                _ => default,
+            };
+
+            if (errorCode != null)
+            {
+                // First look in the analyzer diagnostics of the document. By doing so we get detailed information about the actual
+                // diagnostic message and the code that is affected by the diagnostic id.
+                // If this fails, try to find the error code in the SupportedDiagnostics of all referenced analyzers.
```

Comment:
I think this can be done in a much lesser compute intensive way (not attempting to compute diagnostics and re-using cached information in IDE diagnostic service) and also with much lesser code:

1. You can get to the diagnostic descriptor info cache and host analyzers same as below: https://github.com/dotnet/roslyn/blob/892f9fae716ffb7b6fbd4a32db1eeca8fd376f6f/src/VisualStudio/Core/Def/Implementation/Diagnostics/VisualStudioDiagnosticAnalyzerService.cs#L97-L98
2. Fetch the cached diagnostic descriptors map, grouped by each analyzer file reference by invoking the `GetDiagnosticDescriptorsPerReference`, similar to below, but without the `Transform`: https://github.com/dotnet/roslyn/blob/892f9fae716ffb7b6fbd4a32db1eeca8fd376f6f/src/VisualStudio/Core/Def/Implementation/Diagnostics/VisualStudioDiagnosticAnalyzerService.cs#L119
3. Iterate through all the concatenated diagnostic descriptors in the values of the above returned map, returning the title for the first descriptor that matches the diagnostic ID in the pragma.

Reply:
I think I did something like this by calling `_diagnosticAnalyzerService.GetCachedDiagnosticsAsync` but needed to revert it in b41043e1e8c5c7da79c5fd97060c7ba8a04a18d2 because in the tests that method always returned nothing.

Reply:
`GetCachedDiagnosticsAsync` is different - it relies on already computed diagnostics. Here you are explicitly asking for descriptors, which will be either fetched from the cached or force computed and stored on the cache if not found. Forcing diagnostic computation is much much more expensive operation then just getting all supported analyzer diagnostics.